# Comparison of Mutual Information and DARSI Saliency Map Approaches

This directory contains plots comparing the Mutual Information (MI) approach with the DARSI saliency map approach. These methods are used to generate expression sensitivity plots to mutations, helping identify transcription factor binding sites along the regulatory sequences of *E. coli* operons.

## Dataset
The plots in this directory are based on the MPRA dataset published by Ireland et al., 2020. Each plot corresponds to one of the 95 *E. coli* operons analyzed in the study.

## Plots
Each plot in this directory compares the MI and DARSI approaches for a specific *E. coli* operon. 

### Example Plot
![Example Plot](./xylF.pdf)

**Description:** The plot shows the sensitivity of expression to mutations along the regulatory sequence for *xylF* operon. The top panel displays the gene expression-sensitivity-to-mutation plot generated by DARSI, and the bottom panel displays mutual information between mutations and gene expression at every nucleotide position. Regions with high sensitivity indicate potential transcription factor binding sites. Blue bars are indicative of potential activator binding sites that, when mutated, result in a decrease in expression. Orange bars, on the other hand, represent repressional regions that when mutated result in an increase in gene expression. 

## File Naming Convention
Each file is named after the corresponding operon. For example:
- `acuI.pdf`
- `yqhC.pdf`

## Usage
These plots can be used to:
1. Compare the effectiveness of MI and DARSI approaches.
2. Identify functional regions along regulatory sequences.

